src/main/java/com/formulasearchengine/mathmlsim/similarity/SubTreeComparison.java

```diff
@@ -0,0 +1,174 @@
+     * identical subtree with the comparison tree.
+     *
+     * @param refTree       Reference MathNode tree
+     * @param comTree       Comparison MathNode tree
+     * @param similarities  List of similarities, will be filled during process.
+     * @param holdRefTree   Hold the reference tree in position and only iterate over the comparison tree
+     * @param onlyOperators Find similarities only between operations, no single identifier (end leafs) are checked
+     * @return true - if the current aTree ad bTree are identical subtrees, false otherwise
+     */
+    boolean findSimilarities(MathNode refTree, MathNode comTree, List<Match> similarities, boolean holdRefTree, boolean onlyOperators) {
+        if (isIdenticalTree(refTree, comTree)) {
+            // hit!
+            comTree.setMarked();
+            similarities.add(new Match(refTree, comTree, type));
+            return true;
+        }
+        // iterate the comparison tree over the current node from the ref tree
+        for (MathNode compChild : comTree.getChildren()) {
+            // don't look at leafs if it is already marked
+            // or we only want to compare branching nodes
+            if (compChild.isMarked() || onlyOperators && compChild.isLeaf()) {
+                continue;
+            }
+            // go deeper in the comp. tree but hold the ref tree
+            if (findSimilarities(refTree, compChild, similarities, true, onlyOperators)) {
+                return true;
+            }
+        }
+
+        if (!holdRefTree) {
+            // go deeper in the reference tree
+            for (MathNode refChild : refTree.getChildren()) {
+                if (onlyOperators && refChild.isLeaf()) {
+                    continue;
+                }
+                findSimilarities(refChild, comTree, similarities, false, onlyOperators);
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Are aTree and bTree identical subtrees? If the root node is equal,
+     * all subsequent children will be compared.
+     *
+     * @param aTree first MathNode tree
+     * @param bTree second MathNode tree
+     * @return true - if both trees are identical subtrees, false otherwise
+     */
+    boolean isIdenticalTree(MathNode aTree, MathNode bTree) {
+        // first check if they have the same number of children
+        if (aTree.equals(bTree) && aTree.getChildren().size() == bTree.getChildren().size()) {
+            if (aTree.isOrderSensitive()) {
+                // all children order sensitive
+                for (int i = 0; i < aTree.getChildren().size(); i++) {
+                    if (!isIdenticalTree(aTree.getChildren().get(i), bTree.getChildren().get(i))) {
+                        return false;
+                    }
+                }
+            } else {
+                // order insensitive
+                List<MathNode> bChildren = new ArrayList<>(bTree.getChildren());
```

Comment:
this is somehow complex. Could you add a private subroutine for this type of comparison?